README.md
# AudioPick
A *Chrome Extension* to pick a default audio output device for HTML5 audio and video elements

## How it works
The extension finds HTML5 audio and video elements within the document tree and manipulates the `sinkId` in order
to switch to the desired audio output device. The extension does not and will never find flash objects.
It also does not find audio/video node objects (`new Audio(...)`) that have not been inserted into the document tree.

**Note** that the API requires a successful call to `getUserMedia()` for every site with audio sinks that
need to be manipulated which - as a result - creates entries under `contentSettings['microphone']`, i. e.
it allows those sites to access your microphone.
 
## Links
- [AudioPick Extension on Chrome Webstore](https://chrome.google.com/webstore/detail/audiopick/gfhcppdamigjkficnjnhmnljljhagaha)
- [AudioPick Source on GitHub](https://github.com/rain-fighters/AudioPick)
- [AudioPick Issues on GitHub](https://github.com/rain-fighters/AudioPick/issues)
- [Tech Corner Collection on Google+](https://plus.google.com/collection/swqxgB)
- [Necropola on Google+](https://plus.google.com/108590874920717613332)
- [Necropola on Twitter](https://twitter.com/necropola)
 


CHANGELOG.md
# CHANGELOG

### AudioPick-0.2.2 - 2016-05-21
- going stable
- revert `page_action back` to `browser_action`
- code cleanup

### AudioPick-0.2.1 - 2016-05-18
- quick fix to prevent a loop caused by the *observer*

### AudioPick-0.2.0 - 2016-05-18
- substitute calls to `getUserMedia()` by `get_help_with_GUM`, i. e. write directly to `contentSettings['microphone']` thereby allowing the modification of audio/video on unencrypted pages 
- code cleanup + better diagnostic output
- fixes to handling of asynchronous actions, especially promises
- overwrite devices for an entire frame set, i. e. when sub frames ask the background page for the `default_no`, it asks the top frame and passes this information back to the sub frame
- popup only ever asks the top frame for its current `sink_no` 
 
### AudioPick-0.1.4 - 2016-05-17
- add `'use strict'` to all scripts
- inject the content script into all frames so that `setSinkId()` now also works for embedded audio/video (over HTTPS)
- only call `getUserMedia()` on a site (once) when a call to `setSinkId` actually fails. This should greatly reduce the number of sites added to the list of microphone exceptions.  
- change `browser_action` into `page_action`
- immediately commit changes to the popup device list (commit button removed)
- remember the last temporary `sink_no` choice of a content page when creating the popup device list for it


extension/popup.js
/*
 * $Id: popup.js 57 2016-05-21 19:35:55Z  $
 */
 
'use strict';
 
var bg = chrome.extension.getBackgroundPage();
var default_no = bg.document.getElementById("default_no");
var sink_no = default_no.value;

// -- Update the temporary device selection page 
 function init() {
	log('popup::init');
 	chrome.tabs.query({active: true, currentWindow: true},
		function(tabs) {
			var activeTab = tabs[0];
			log('popup::init Active tab: ' + activeTab.url || '[NO URL FOUND]')
			log('popup::init Sending message: report_sink_no');
			chrome.tabs.sendMessage(activeTab.id,
				{"message": "report_sink_no"},
				{'frameId': 0}, // only request from main frame
				function(response) {
					if (response) {
						log("popup::init::msg Received Response: " + response.sink_no);
						sink_no = response.sink_no;
					}
					navigator.mediaDevices.enumerateDevices()
						.then(update_device_popup)
						.catch(errorCallback);
				}
			);
		}
	)
}

function errorCallback(error) {
	log('error: ', error);
}

function log(message) {
	// logging to background console
	bg.console.log(
		'%cpopup:%c ' +  message,
		'border-radius: 0.15em; background: #FFCC4480; font-weight: 600;'
		+ 'padding: 0.15em 0.4em; ', ''
	)
}

function update_device_popup(deviceInfos) {
	log('update_device_popup: ' + deviceInfos.length + ' device(s) total (audio/video input/output)');
	var div = document.getElementById("device_options");
	var select = bg.document.getElementById("device_cache");
	while (div.firstChild) { div.removeChild(div.firstChild); }
	for (var i = 0; i !== deviceInfos.length; ++i) {
		var kind = deviceInfos[i].kind;
		var id = deviceInfos[i].deviceId;
		var text = deviceInfos[i].label;
		log('update_device_popup::device\n\t' + id + ' - ' + text);
		if (kind === 'audiooutput') {
			if (id == "default") {
				text = "System Default Device";
			} else if (id == "communications") {
				text = "System Default Communications Device";
			}
			var option = bg.document.getElementById(id);
			if (!text) {
				if (option && option.value) {
					text = option.value;
				} else {
					text = id;
				}
			}
			if (option) {
				option.value = text;
			} else {
				option = bg.document.createElement("option");
				option.id = id;
				option.value = text;
				select.appendChild(option);				
			}
			var input = document.createElement("input");
			input.type= "radio";
			input.name = "device";
			input.id = id;
			input.value = i;
			input.onchange = function(e){input_onchange(e);};
			var textNode = document.createTextNode(text);
			var label = document.createElement("label");
			if (i == sink_no) {
				log('update_device_popup::current_default_no: ' + i || '[NO SINK_NO]' + ' - ' + id || '[NO ID]' + ' - ' + text);
				input.checked = true;
			}			
			label.appendChild(textNode);
			label.appendChild(input);
			div.appendChild(label);
		}
	}
}

function update_default_no(e) {
	log("update_default_no: " +  e.target.value);
	var default_no =  bg.document.getElementById("default_no");
	default_no.value = e.target.value;
	chrome.storage.local.set({"AP_default_no" : e.target.value});
}

function input_onchange(e) {
	log('popup::input_onchange::browser_action Commit');
	var sink_no = e.target.value;	

	// Override per page logic for setting default
	log('popup::input_onchange::set_default OVERRIDE')
	log('popup::input_onchange::set_default Updating default:\n  Id: ' 
	    + e.target.id + '\nValue: ' + e.target.value)
	update_default_no(e)
	log('popup::closing')
	window.close();

	return 
	chrome.tabs.query({active: true, currentWindow: true},
		function(tabs) {
			var activeTab = tabs[0];
			log('popup::input_onchange Sending message: browser_action_commit, sink_no: ' + sink_no);
			chrome.tabs.sendMessage(activeTab.id, { // send to all frames without using options = {'frameId': N} 
				"message": "browser_action_commit",
				"sink_no":  sink_no
			});
			window.close();
		}
	);
};

// -- main ---------------------------------------------------------------
init();



extension/content.js
/*
 * $Id: content.js 57 2016-05-21 19:35:55Z  $
 */

 'use strict';
 
var sink_no = 0;
var sink_name = 'System Default Device';
var sink_id = 'default';
var frame_url = location.protocol + '//'+ location.host + location.pathname;
var frame_depth = get_depth(window.self);
var GUM_state = undefined;
/*
 * undefined == wait with getUserMedia() until we know that we actually need to call setSinkId 
 *         0 == last call to setSinkId failed. Going to call getUserMedia() next time
 *         1 == getUserMedia() succeeded 
 *        -1 == getUserMedia() failed 
 */
 
function get_depth(w) {
	if (w.parent == w) {
		return 0;
	} else {
		return 1 + get_depth(w.parent);
	}
}
function log(message) {
	console.log('  '.repeat(frame_depth) + 'AudioPick(' + frame_url + '): ' + message);
}

function errorCallback(error) {
	log('Error: '+ error);
}

// -- Register a listener for messages from the popup page
function register_message_listener() {
	chrome.runtime.onMessage.addListener(
		function(request, sender, sendResponse) {
			if (request.message === "browser_action_commit" ) {
				log('Received message: browser_action_commit, sink_no: ' + request.sink_no + ', sink_name: ' + request.sink_name);
				if (request.sink_no != undefined) {
					sink_no = request.sink_no;
					sink_name = request.sink_name;
					get_devices(); // --> inspect_device_infos() --> update_all_sinks()
				}
			} else if (request.message == "report_sink_no") {
				log('Received message: report_sink_no');
				log('Reply with: sink_no: ' + sink_no + ', sink_name: ' + sink_name);
				sendResponse({'sink_no': sink_no, 'sink_name': sink_name});
			}
		}
	)
}	

// -- Register a Mutation Observer to monitor changes and additions of <audio/> and <video/> elements
function register_observer() {
	var observer = new MutationObserver(
		function(mutations) {
			var needs_update = false;
			mutations.forEach(
				function(mutation) {
					//log('mutation.type: ' + mutation.type);
					if (mutation.type == 'attributes') {
						// This can cause a loop!
						//if (check_node(mutation.target)) needs_update = true;
					} else {
						for (var i = 0; i < mutation.addedNodes.length; i++) {
							if (check_node(mutation.addedNodes[i])) needs_update = true;
						}
					}
				}
			);
			if (needs_update) update_all_sinks();
		}
	);
	observer.observe(document.documentElement, {
		childList: true,
		subtree: true,
		attributes: true,
		characterData: true
	});
}

function check_node(node) {
	var name = node.nodeName;
	var attributes = node.attributes
	if ((name == 'AUDIO') || (name == 'VIDEO')) {
		log('node added/changed: ' + name);
		return true;
	}
	return false;
}

// -- Request the default sink_no from the background
function request_default_no() {
	log('Requesting default_no ...');
	chrome.runtime.sendMessage({'method': 'AP_get_default_no'},
		function(response) {
			if (response) {
				log('Received default_no: ' + response.default_no + ' default_name: ' + response.default_name);
				sink_no = response.default_no;
				sink_name = response.default_name;
				get_devices();
			}
		}
	)
}

function request_help_with_GUM() {
	chrome.runtime.sendMessage({'method': 'AP_help_with_GUM', 'primaryPattern': location.protocol + '//'+ location.host + '/*'},
		function(response) {
			if (response) {
				log('Received result: ' + response.result);
				GUM_state = 1;
				update_all_sinks();
			}
		}
	);
}

function get_devices() {
	navigator.mediaDevices.enumerateDevices()
		.then(inspect_devices)
		.catch(errorCallback);
}

function inspect_devices(deviceInfos) {
	let deviceInfo;
	log('Inspecting Devices: ' + deviceInfos.length + ' device(s) total (audio/video input/output)');
	let found = false;
	for (var i = 0; i != deviceInfos.length; i++) {
		deviceInfo = deviceInfos[i];
		//log('  Devices[' + i + ']: ' + deviceInfo.kind + ': ' + deviceInfo.deviceId);
		if (deviceInfo.kind == 'audiooutput' && deviceInfo.label == sink_name) {
			log('Selecting Devices[' + i + ']: ' + deviceInfo.label + " (" + deviceInfo.deviceId + ")");
			sink_id = deviceInfo.deviceId;
			sink_no = i;
			found = true;
			break;
		}
	}
	if (!found) {
		for (var i = 0; i != deviceInfos.length; i++) {
			deviceInfo = deviceInfos[i];
			//log('  Devices[' + i + ']: ' + deviceInfo.kind + ': ' + deviceInfo.deviceId);
			if (deviceInfo.kind == 'audiooutput') {
				log(sink_name + ' not found! Selecting Devices[' + i + ']: ' + deviceInfo.label + " (" + deviceInfo.deviceId + ")");
				sink_id = deviceInfo.deviceId;
				sink_no = i;
				sink_name = deviceInfo.label;
				break;
			}
		}
	}
	with_or_without_GUM();
}

function with_or_without_GUM() {
	if (GUM_state == 0) {
		request_help_with_GUM();
	} else {
		update_all_sinks();		
	}	
}

function update_all_sinks() {
	var promises = [];
	var allMedia = document.querySelectorAll('audio,video');
	for (var j = 0; j < allMedia.length; j++) {
		var name = allMedia[j].nodeName;
		var src = allMedia[j].currentSrc;
		log('  Queuing SetSinkId: ' + j + ': ' + name + ': ' +  src + ': ' + sink_id);
		promises.push(allMedia[j].setSinkId(sink_id));
	}
	if (promises.length > 0) {
		log('Tyring to update all (' + promises.length + ') sinks (GUM_state == ' + GUM_state + '): ' + sink_id);
		Promise.all(promises)
			.then(function(results){log('All set.'); })
			.catch(function(error){
				if (GUM_state == undefined) {
					GUM_state = 0;
					log('SetSinkId failed: ' + error + '. Retrying with GUM ...');
					with_or_without_GUM();
				} else {
					log('SetSinkId failed: ' + error + '.  Giving up.');
				}
			});
	} else {
		log('No sinks found');
	}
	register_observer();
}

// -- main ---------------------------------------------------------------
register_message_listener();
request_default_no();




extension/options.js
/*
 * $Id: options.js 36 2016-05-15 23:06:46Z  $
 */
 
var bg = chrome.extension.getBackgroundPage();
 
// -- Help the background page in retrieving devices and update the default device selection page 
 function init() {
 	// We could include the around 2500 lines of code from WebRTCs "adapter.js"
	// and then call "navigator.mediaDevices.getUserMedia()", but why should we?
	navigator.webkitGetUserMedia(
		{
			audio:true,
			video:false
		},
		function(stream) {
			log('getUserMedia: success (as it should be, when called from options full page)');
			navigator.mediaDevices.enumerateDevices()
				.then(update_device_options)
				.catch(errorCallback);
		},
		function(stream) {
			log('getUserMedia: error (as is expected, when run from options_ui)');
			navigator.mediaDevices.enumerateDevices()
				// Inspecting enumerated devices (that look fucky)
				.then((devices) => {
					devices.map(device => Object.fromEntries(Object.entries(device)))
						   .forEach(device => log('Enumerated device log:\n' + JSON.stringify(device, null, 4)) )
					update_device_options(devices)
				})
				.catch((e) => {
					log('init::getUserMedia error during enumerateDevices() after error in getUserMedia:\n\t"' + error + '"')
				})
				// .catch(errorCallback);
		}
	);
}

function errorCallback(error) {
	log('error: '+ error);
}

function log(message) {
	// logging to background console
	bg.console.log(
		'%coptions:%c ' +  message,
		'border-radius: 0.15em; background: #55FF2240; font-weight: 600;'
		+ 'padding: 0.15em 0.4em; ', ''
	)
}

function update_default_no(e) {
	log("update_default_no: " +  e.target.value);
	var default_no =  bg.document.getElementById("default_no");
	default_no.value = e.target.value;
	chrome.storage.local.set({"AP_default_no" : e.target.value});
}

/**
 * 
 * @param {MediaDeviceInfo[]} deviceInfos 
 */
function update_device_options(deviceInfos) {
	log('update_device_options: ' + deviceInfos.length + ' device(s) total (audio/video input/output)');
	var div = document.getElementById("device_options");
	var select = bg.document.getElementById("device_cache");

	log('update_device_options: ' + select.children.length + ' cached devices found before updating.')

	var default_no = bg.document.getElementById("default_no");
	while (div.firstChild) { div.removeChild(div.firstChild); }
	for (var i = 0; i !== deviceInfos.length; ++i) {
		var kind = deviceInfos[i].kind;
		var id = deviceInfos[i].deviceId;
		var text = deviceInfos[i].label;

		// Check if device has actual id and value before polluting DOM
		if(!(id && text))
		{
			log('update_device_options::add_device ABORTING ! Missing id or value (id: "' + (id || '') + '" value: "' + (value || '') + '")')
			continue
		}

		log('update_device_options::add_device Reviewing device:\n\t' + id || '[NO ID]' + ' - ' + text);
		if (kind === 'audiooutput') {
			if (id == "default") {
				text = "System Default Device";
			} else if (id == "communications") {
				text = "System Default Communications Device";
			}
			var option = bg.document.getElementById(id);
			if (!text) {
				if (option && option.value) {
					text = option.value;
				} else {
					text = id;
				}
			}
			if (option) {
				option.value = text;
			} else {
				option = bg.document.createElement("option");
				option.id = id;
				option.value = text;
				select.appendChild(option);				
			}
			var input = document.createElement("input");
			input.type= "radio";
			input.name = "device";
			input.id = id;
			input.value = i;
			input.onchange = function(e){update_default_no(e);};
			var textNode = document.createTextNode(text);
			var label = document.createElement("label");
			if (i == default_no.value) {
				log('update_device_options::current_default_no ' + i + ' - ' + id + ' - ' + text);
				input.checked = true;
			}			
			label.appendChild(input);
			label.appendChild(textNode);
			div.appendChild(label);
		}
	}
}

// -- main ---------------------------------------------------------------
init();


extension/background.js
/*
 * $Id: background.js 57 2016-05-21 19:35:55Z  $
 */

 'use strict';
  
 var stored_id = 'default'
 var stored_no = 0;
 var extension_id = chrome.runtime.id;
 
 // Haha, it seems we no longer need getUserMedia() ...
chrome.contentSettings['microphone'].set({'primaryPattern':'*://' + extension_id + '/*','setting':'allow'});

chrome.runtime.onMessage.addListener(
	function (message, sender, sendResponse) {
		if (message.method == "AP_get_default_no") {
			log('Received message: ' + message.method + ' from frame ' + sender.frameId + ' on tab ' + sender.tab.id);
			if (sender.frameId != 0 ) {
				log('Asking top frame: report_sink_no');
				chrome.tabs.sendMessage(sender.tab.id,
					{"message": "report_sink_no"},
					{'frameId': 0},  // request from top frame
					function(response) {
						if (response) {
							var default_no = document.getElementById("default_no");
							log("Received Response from top frame: " + response.sink_no);
							if (response.sink_no != 0) {
								log('Reply to sub frame ' + sender.frameId + ' with: top sink_no: ' + response.sink_no);
								sendResponse({'default_no': response.sink_no});
							} else {
								log('Reply to sub frame ' + sender.frameId + ' with: default_no: ' + default_no.value);
								sendResponse({'default_no': default_no.value});
							}
						}
					}
				);
			} else {
				var default_no = document.getElementById("default_no");
				log('Reply with: default_no: ' + default_no.value);
				sendResponse({'default_no': default_no.value});
			}
		} else if (message.method == "AP_help_with_GUM") {
			log('Received message: ' + message.method + ', primaryPattern: ' + message.primaryPattern);
			chrome.contentSettings['microphone'].set({'primaryPattern': message.primaryPattern,'setting':'allow'});
			log('Reply with: result: ' + 'Have fun!');
			sendResponse({'result': 'Have fun!'});
		}
		return true;
    }
 )
 
// -- Initialize device_cache (list of available devices)
function init() {
	var default_no = document.getElementById("default_no");
	default_no.value = stored_no;
	navigator.mediaDevices.enumerateDevices()
		.then(update_device_cache)		
		.catch((e) => {
			log('[ERROR!] Failed enumerateDevices during background init device_cache initialization.')
		})
}
function errorCallback(error) {
	log('error: '+ error);
}

function log(message) {
	console.log(
		'%cbackground:%c ' +  message,
		'border-radius: 0.15em; background: #55AAFF40; font-weight: 600;'
		+ 'padding: 0.15em 0.4em; ', ''
	);
}
                                                                                                                                                                                                                                                                                                           
function update_device_cache(deviceInfos) {
	var default_no = document.getElementById("default_no");
	var select = document.getElementById('device_cache');
	log('update_device_cache: ' + deviceInfos.length + ' device(s) total (audio/video input/output)');
	for (var i = 0; i !== deviceInfos.length; ++i) {
		var kind = deviceInfos[i].kind;
		var id = deviceInfos[i].deviceId;
		var text = deviceInfos[i].label;
		log('update_device_cache::device:\n\t' + id + ' - ' + text);
		if (kind === 'audiooutput') {
			if (id == "default") {
				if (stored_no == 0) {
					stored_no = i;
					default_no.value = stored_no;
				}
				text = "System Default Device";
			} else if (id == "communications") {
				text = "System Default Communications Device";
			}
			log('update_device_cache::device::audiooutput:\n\t' + id + ' - ' + text);
			if (text) { // only update/write cache, when we have a device label
				var option = document.getElementById(id)
				if (option) {
					option.value = text;
				} else {
					option = document.createElement("option");
					option.id = id;
					option.value = text;
					select.appendChild(option);				
				}
			}
		}
	}
}

// -- main ---------------------------------------------------------------
chrome.storage.local.get("AP_default_no",
	function(result) {
		stored_no = result["AP_default_no"];
		if (!stored_no) { stored_no = 0; }
		log('stored_no: '+ stored_no);
		init();
	}
);
